### PR TITLE
Jsonified

### PR DESCRIPTION
**Pull request by turbocrime**
instead of aliasing jsonvalue, actually derive json-compatible type information

use it like:

```ts
const x = new AnyMessage({});
const xJson = x.toJson() as Jsonified<AnyMessage>;
```

### CODE REVIEW

File: packages/types/src/jsonified.ts
Comment (turbocrime) [inline packages/types/src/jsonified.ts:1]: jsonified.ts is the important part

currently, putting bufbuild message types in here requires wrapping in `PlainMessage<T>` first, maybe we could add support for doing that automatically

**Comment by turbocrime** [comment]
minimized for @jessepinho use, but more wip available

File: packages/types/src/jsonified.ts
Comment (turbocrime) [reply packages/types/src/jsonified.ts:1]: ok, added that

File: packages/types/src/jsonified.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+      [P in keyof T as T[P] extends Function ? never : P]: JsonField<T[P]>;
+    };
+
+// prettier-ignore
+type JsonField<F> = F extends (JsonValue) ? F
+                  : F extends (Date| Uint8Array | bigint) ? string
+                  : F extends (infer U)[] ? JsonField<U>[]
+                  : F extends Message<infer U> ? Jsonified<PlainMessage<U>>
+                  : F extends object ? Jsonified<F>
+                  : F;
```
Comment (jessepinho) [inline packages/types/src/jsonified.ts]: It's... it's beautiful 🥺

File: packages/types/src/indexed-db.ts [hunk 2]
```diff
@@ -93,27 +91,27 @@ export interface PenumbraDb extends DBSchema {
   // ======= Json serialized values =======
   // Allows wasm crate to directly deserialize
   ASSETS: {
-    key: Base64Str; // Jsonified<DenomMetadata['penumbraAssetId']['inner']>
-    value: Jsonified<DenomMetadata>;
+    key: Base64Str; // DenomMetadata['penumbraAssetId']['inner']
+    value: JsonValue;
```
Comment (jessepinho) [inline packages/types/src/indexed-db.ts:95]: Why change these to `JsonValue`? Isn't e.g., `Jsonified<DenomMetadata>` more useful now that we have a proper `Jsonified` type?

File: apps/extension/src/wasm-build-action.ts [hunk 3]
```diff
@@ -43,7 +42,7 @@ async function executeWorker(
   witness: WitnessData,
   fullViewingKey: string,
   actionPlanIndex: number,
-): Promise<Jsonified<Action>> {
+): Promise<JsonValue> {
```
Comment (jessepinho) [inline apps/extension/src/wasm-build-action.ts:45]: Same Q here — don't we want `Jsonified<Action>`?

File: apps/extension/src/offscreen.ts [hunk 4]
```diff
@@ -49,16 +47,16 @@ export const buildActionHandler = (request: ActionBuildRequest) => {
...
   fullViewingKey: string,
   actionPlanIndex: number,
-): Promise<Jsonified<Action>> => {
+): Promise<JsonValue> => {
   return new Promise((resolve, reject) => {
     const worker = new Worker(new URL('./wasm-build-action.ts', import.meta.url));
 
     const onWorkerMessage = (e: MessageEvent) => {
-      resolve(e.data as Jsonified<Action>);
+      resolve(e.data as JsonValue);
```
Comment (jessepinho) [inline apps/extension/src/offscreen.ts:59]: Same Q here

**Review by jessepinho — COMMENTED** [review]
Not sure I understand the changes to using `JsonValue` instead of `Jsonified<X>` — could you please explain this change before I approve? Otherwise, the `Jsonified` type looks 🔥

File: packages/types/src/indexed-db.ts (hunk 2 above)
Comment (turbocrime) [reply packages/types/src/indexed-db.ts:95]: the upgrade has complex downstream effects :) for now it's just stubbed out

File: apps/extension/src/offscreen.ts (hunk 4 above)
Comment (turbocrime) [reply apps/extension/src/offscreen.ts:59]: events arriving here are technically 'unknown' and must be inspected with a type guard

File: apps/extension/src/wasm-build-action.ts (hunk 3 above)
Comment (turbocrime) [reply apps/extension/src/wasm-build-action.ts:45]: we could probably assert that, but at the return statement we're calling a function that really does return `JsonValue` explicitly

**Review by jessepinho — APPROVED** [review]
OK, LGTM! Looks like there are some build/lint issues though — once those are green, ship it 🚢

**Review by grod220 — COMMENTED** [review]
I'm a bit on the fence with this PR. Few things:

- We really shouldn't be interacting with json-serialized-object fields directly. We almost should pretend they are bytes. It's really the deserialized classes that are the first class citizens.
- The new `Jsonified` type may prove to be difficult to maintain.
- It doesn't appear that JsonValue === Jsonified<T> anymore. This is evident by the fact we can't use it to type the indexeddb values anymore.

Given this, I don't see a big benefit with expanding the type. Am I missing something?

**Comment by jessepinho** [comment]
Not sure what you mean by "expanding the type," but the idea here is that `Jsonified<X>` is more useful than `JsonValue` since we know what shape it takes. That said, if we shouldn't be interacting with JSON-serialized objects directly, I suppose it's not that useful except as documentation of what object type it's serialized from.

**Pull request merged**